# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [27]:
value_children = data.groupby('children')['debt'].count()
value_children

children
0    14091
1     4808
2     2052
3      330
4       41
5        9
Name: debt, dtype: int64

In [28]:
value_children_debt = data.groupby('children')['debt'].sum()
value_children_debt

children
0    1063
1     444
2     194
3      27
4       4
5       0
Name: debt, dtype: int64

In [29]:
children_debt_procent = value_children_debt/value_children*100
children_debt_procent

children
0    7.543822
1    9.234609
2    9.454191
3    8.181818
4    9.756098
5    0.000000
Name: debt, dtype: float64

**Вывод:** Прямой зависимости невозврата кредитов в срок от увеличения количества детей у заёмщиков не обнаружено. И всё же среди кредитополучателей, не имеющих детей, вероятность выплаты кредита в срок выше, чем у кредитополучателей с детьми. 

В ходе анализа выявлено:
1. У заёмщиков без детей процент невозврата в срок ниже - 7,54%.
2. У заёмщиков с 1 и 2 детьми процент невозврата отличается незначительно - 9.23% и 9,45% соответственно. 
3. У заёмщиков с 3 и 4 детьми процент невозврата 8,18% и 9,75% соответственно. 
4. У заёмщиков с 5 детьми - 0% невозврата в срок.
5. Значения процентов невозврата для заёмщиков с 3, 4 или 5 детьми считаю некорретными из-за слишком маленькой выборки заёмщиков для анализа.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [ ]:
value_family_debt = data.groupby('family_status')['debt'].sum()
value_family_debt.sort_values()

In [ ]:
value_family = data.groupby('family_status')['debt'].count()
value_family.sort_values()

In [ ]:
debt_family_persent = value_family_debt/value_family*100
debt_family_persent.sort_values()

**Вывод:** Анализ показал, что между семейным положением и невозвратом в срок есть зависимость.
1. У вдовцов/вдов процент невозврата в срок меньше всех - 6,62%.
2. У заёмщиков в разводе и женатых/замужних заёмщиков процент невозврата чуть больше - 7,06% и 7,56% соответственно.
3. У неженатых/незамужних заёмщиков самый большой процент невозврата в срок 9,31% и 9,76% соответственно.

<div class="alert alert-success">
<b>ОТЛИЧНО! 👍</b>

Здесь также верно. Кредитному отделу следует с осторожностью относиться к клиентам, которые не состоят или не состояли в браке. Согласно данным такие клиенты менее ответственные. Получается, что узаконенные отношения ведут к более ответственному отношению к своим обязательствам. Ну вроде как совпадает со здравым смыслом - семейные люди более скрупулёзно берегут свой семейное спокойствие и несут за него ответственность.
</div>


#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [31]:
total_income_debt = data.groupby('total_income_category')['debt'].sum()
total_income_debt

total_income_category
A       2
B     354
C    1353
D      21
E       2
Name: debt, dtype: int64

In [32]:
total_income = data.groupby('total_income_category')['debt'].count()
total_income

total_income_category
A       25
B     5014
C    15921
D      349
E       22
Name: debt, dtype: int64

In [33]:
debt_total_income_persent = total_income_debt/total_income*100
debt_total_income_persent

total_income_category
A    8.000000
B    7.060231
C    8.498210
D    6.017192
E    9.090909
Name: debt, dtype: float64

In [ ]:
debt_total_income_persent.sort_values()

**Вывод:** Процент невозврата кредита в срок с увеличением дохода не уменьшается. Значит мы не можем сделать вывод, что чем больше доход заёмщика, тем вероятнее, что он вернёт в срок. 
1. Самый низкий процент невозврата в срок у заёмщиков с доходами от 30 001 до 50 000 (категория D) - 6.02%.
2. Самый высокий процент невозврата в срок у заёмщиков с доходами от 30 000 и ниже (категория E) - 9.09%.
3. У заёмщиков с доходами от 200 001 до 1 000 000 (категория В) процент невозврата в срок немного выше, чем у категории D - 7.06%.
4. У заёмщиков с доходами от 1 000 001 и выше (категория А) и с доходами от 50 001 до 200 000 (категория С) проценты невозврата в срок - 8% и 8,5% соответственно.
5. Выводы относительно категорий А и Е ставлю под сомнение, т.к. выборка крайне недостаточна для однозначных выводов.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [ ]:
purpose_debt = data.groupby('purpose_category')['debt'].sum()
purpose_debt.sort_values()

In [ ]:
purpose_debt_value = data.groupby('purpose_category')['debt'].count()
purpose_debt_value.sort_values()

In [ ]:
purpose_debt_percent = purpose_debt/purpose_debt_value*100
purpose_debt_percent.sort_values()

**Вывод:** 
1. У кредитов, взятых на надвижимость, самый меньший процент невозврата в срок - 7.3%.
2. У кредитов, взятых на проведение свадьбы, процент невозврата в срок чуть больше, чем у кредитов на недвижимость - 7.91%.
3. У кредитов, взятых на получение образования и на операции с автомобилем, самые большие проценты невозврата в срок - 9.25% и 9.35% соответственно.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* возможная причина появления пропусков в столбце 'days_employed' - технический сбой. 
На счёт пропусков в столбце 'total_income' есть сомнения: возможно, и человеческий фактор (не указали доход), и технический сбой.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Заполнить данные средними значениями было бы некорректно - данные сильно бы изменились в сторону увеличения.

### Шаг 4: общий вывод.

Было проведено исследование с целью выявления влияния различных факторов на погашение кредита в срок. Входные данные - статистика о платёжеспообности клиентов. Результаты исследования будут полезны при построении модели кредитного скоринга - специальной программы, которая оценивает способность потенциального клиента вернуть кредит банку в срок.

Перед проведением исследования были сформулированы следующие гипотезы:
- Гипотеза 1 - Количество детей влияет на погашение кредита в срок.
- Гипотеза 2 - Семейное положение влияет на погашение кредита в срок.
- Гипотеза 3 - Уровень дохода влияет на погашение кредита в срок.
- Гипотеза 4 - Цель кредита влияет на погашение кредита в срок.

В результате удалось выявить определенные зависимости.

**Гипотеза 1 частично подтверждена:** среди кредитополучателей, не имеющих детей, вероятность выплаты кредита в срок выше, чем у кредитополучателей с детьми.

- У заёмщиков без детей процент невозврата кредита в срок - 7,54%.
- У заёмщиков с 1 и 2 детьми проценты невозврата кредита в срок - 9.23% и 9,45% соответственно.

Рекомендация: провести дополнительное исследование на большей выборке данных, т.к. для заёмщиков с 3, 4 и 5 детьми выводы неочевидны.

**Гипотеза 2 подтверждена:** у заёмщиков, состоявших в браке или находящихся в браке на текущий момент, вероятность погашения кредита в срок выше, чем у неженатых/незамужних заёмщиков.

- У вдовцов/вдов процент невозврата в срок меньше всех - 6,62%.
- У заёмщиков в разводе и женатых/замужних заёмщиков процент невозврата чуть больше - 7,06% и 7,56% соответственно.
- У неженатых/незамужних заёмщиков самый большой процент невозврата в срок - 9,31% и 9,76% соответственно.

**Гипотеза 3 не подтверждена:** уровень дохода не влияет на способность клиента вернуть заёмные средства в срок. 

- Самый низкий процент невозврата в срок у заёмщиков с доходами от 30 001 до 50 000 (категория D) - 6.02%.
- У заёмщиков с доходами от 200 001 до 1 000 000 (категория В) процент невозврата в срок немного выше, чем у категории D - 7.06%.
- У заёмщиков с доходами от 50 001 до 200 000 (категория С) проценты невозврата в срок - 8,5%.

Рекомендация: провести дополнительное исследование на большей выборке данных, т.к. для заёмщиков категорий А и Е выводы неочевидны.

**Гипотеза 4 подтверждена:** у заёмщиков, берущих кредит на недвижимость или на проведение свадьбы, вероятность погашения кредита в срок выше, чем у заёмщиков, берущих кредит на получение образования или на операции с автомобилем.

- У кредитов, взятых на надвижимость, самый меньший процент невозврата в срок - 7.3%.
- У кредитов, взятых на проведение свадьбы, процент невозврата в срок чуть больше, чем у кредитов на недвижимость - 7.91%.
- У кредитов, взятых на получение образования и на операции с автомобилем, самые большие проценты невозврата в срок - 9.25% и 9.35% соответственно.

**Общие рекомендации:**
1. Специалистам по моделированию кредитного скоринга рекомендую использовать зависимости из 2 и 4 гипотез (о семейном положении и о цели кредита).

2. Необходимо собрать более полную выборку для проведения повторной проверки 1 и 3 гипотез (о детях заёмщиков и о уровне дохода).